packages/visual-editor/src/components/Dropzone/Hitboxes.tsx

```diff
@@ -0,0 +1,206 @@
+import React, { CSSProperties, useCallback, useEffect, useMemo, useState } from 'react';
+import styles from './styles.module.css';
+import { useTreeStore } from '@/store/tree';
+import { getItemDepthFromTree } from '@/utils/getItem';
+import { CTFL_ZONE_ID, DRAGGABLE_WIDTH, ROOT_ID } from '@/types/constants';
+import { useZoneStore } from '@/store/zone';
+import { useDraggedItemStore } from '@/store/draggedItem';
+import { createPortal } from 'react-dom';
+
+interface Props {
+  parentZoneId: string;
+  zoneId: string;
+  enableRootHitboxes: boolean;
+}
+
+const INITIAL_OFFSET = 10;
+const OFFSET_INCREMENT = 8;
+const HITBOX_HEIGHT = 20;
+const HITBOX_WIDTH = 80;
+const MIN_HEIGHT = 45;
+const MIN_DEPTH = 20;
+enum HitboxDirection {
+  TOP,
+  LEFT,
+  RIGHT,
+  BOTTOM,
+  SELF_VERTICAL,
+  SELF_HORIZONTAL,
+}
+
+const calcOffsetDepth = (depth: number) => {
+  return INITIAL_OFFSET - OFFSET_INCREMENT * depth;
+};
+
+const Hitboxes: React.FC<Props> = ({ zoneId, parentZoneId, enableRootHitboxes }) => {
+  const tree = useTreeStore((state) => state.tree);
+  const isDraggingOnCanvas = useDraggedItemStore((state) => state.isDraggingOnCanvas);
+  const zoneDepth = useMemo(
+    () => getItemDepthFromTree({ id: parentZoneId }, tree),
+    [tree, parentZoneId],
+  );
+  const [fetchDomRect, setFetchDomRect] = useState(Date.now());
+
+  useEffect(() => {
+    /**
+     * A bit hacky but we need to wait a very small amount
+     * of time to fetch the dom getBoundingClientRect once a
+     * drag starts because we need pre-drag styles like padding
+     * applied before we calculate positions of hitboxes
+     */
+    setTimeout(() => {
+      setFetchDomRect(Date.now());
+    }, 50);
+  }, [isDraggingOnCanvas]);
+
+  const hitboxContainer = useMemo(() => {
+    return document.querySelector('[data-ctfl-hitboxes]');
+  }, []);
+
+  const domRect = useMemo(() => {
+    return document.querySelector(`[${CTFL_ZONE_ID}="${zoneId}"]`)?.getBoundingClientRect();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [zoneId, fetchDomRect]);
+
+  const zones = useZoneStore((state) => state.zones);
+
+  const direction = zones[parentZoneId]?.direction || 'vertical';
+  const isVertical = direction === 'vertical';
+
+  const isRoot = parentZoneId === ROOT_ID;
+
+  const getStyles = useCallback(
```

Comment:
Maybe we can move this code to a different file - either a styles file or some utils.ts file?